encoding/ccf/decode.go

```diff
@@ -345,46 +358,56 @@ func (d *Decoder) decodeValue(t cadence.Type, types *cadenceTypeByCCFTypeID) (ca
 		return d.decodeUFix64()
 
 	case *cadence.VariableSizedArrayType:
-		return d.decodeArray(typ, false, 0, types)
+		return d.decodeArray(t, false, 0, types)
 
 	case *cadence.ConstantSizedArrayType:
-		return d.decodeArray(typ, true, uint64(typ.Size), types)
+		return d.decodeArray(t, true, uint64(t.Size), types)
 
 	case *cadence.DictionaryType:
-		return d.decodeDictionary(typ, types)
+		return d.decodeDictionary(t, types)
 
 	case *cadence.ResourceType:
-		return d.decodeResource(typ, types)
+		return d.decodeResource(t, types)
 
 	case *cadence.StructType:
-		return d.decodeStruct(typ, types)
+		return d.decodeStruct(t, types)
 
 	case *cadence.EventType:
-		return d.decodeEvent(typ, types)
+		return d.decodeEvent(t, types)
 
 	case *cadence.ContractType:
-		return d.decodeContract(typ, types)
+		return d.decodeContract(t, types)
+
+	case cadence.StoragePathType:
+		return d.decodePath()
+
+	case cadence.PublicPathType:
+		return d.decodePath()
 
-	case cadence.PathType:
+	case cadence.PrivatePathType:
 		return d.decodePath()
 
 	case cadence.MetaType:
-		return d.decodeNullableTypeValue()
+		typeValue, err := d.decodeNullableTypeValue(newCadenceTypeByCCFTypeID())
```

Comment:
Can we pass `types` instead of constructing a new `cadenceTypeByCCFTypeID`?

Reply:
Unfortunately, no.  

`cadenceTypeByCCFTypeID` uses a map with CCF type ID as keys.  CCF type ID can collide if we reuse type definition's `types` for type values because:
- CCF type IDs are zero-based for composite `TypeValue`.
- CCF type IDs are zero-based for composite type definitions.

Reply:
Ah, right. I had the same question for encoding I think. 

Could you maybe add this explanation as a comment?

Reply:
Done in commit 4a6aa577f9b69074f1dfc0e20743031e73e1486a.